articulate/paperplane
# v2

![dilbert](http://www.realclear.com/comics/dilbert/images/2016/01/dilbert_january_15_2016.gif)

I thought about titling this PR **"Hurray for ADT's!"**, because  that's basically what it's all about.  I heard that `one of our backend projects` was using `paperplane` with ADT's, and in the process was having to wrap all the bits in converter functions to make them work, and... I felt bad.  I mean, that's what `paperplane` is supposed to be about: making backend functional programming easy.  

So, this PR is me fixing that, and making other changes I've been wanting to make, too.

## What's new/changing:
- Support for request handlers that return ADT's (not just Promises) via Redux middleware.
- `mount` is becoming a unary function that accepts a single options object.
- All options have sane defaults (like using `paperplane.logger` as the default `logger`).
- Much simpler error handling via the new `cry` option (and new [`paperplane-airbrake` to match](https://github.com/articulate/paperplane-airbrake/pull/5)).
- Lots of internal cleanup to use other tools that we've extracted in the past year.

For details on the breaking changes, please read the [Migration guide](https://github.com/articulate/paperplane/blob/v2/docs/migration-guide.md).

## How to test:
- [ ] Checkout this branch and https://github.com/articulate/paperplane-airbrake/pull/5 locally.
- [ ] Pick your favorite existing project that already uses `paperplane`.
- [ ] Volume-in the new versions by adding the following to your `docker-compose.override.yml`:
```yml
volumes:
  - '../paperplane:/service/node_modules/paperplane'
  - '../paperplane-airbrake:/service/node_modules/paperplane-airbrake'
```
- [ ] Follow all of the instructions in the [Migration guide](https://github.com/articulate/paperplane/blob/v2/docs/migration-guide.md) to convert your app to v2.
- [ ] Check and make sure everything still works.
- [ ] Then make sure your `APP_ENV=dev`, but set `NODE_ENV=production`.
- [ ] Throw some errors somewhere, and make sure they appear in Airbrake with the request data.
- [ ] Read the updated docs and decide if they make sense.
- [ ] Invite your friends and coworkers to also do the same.

**Special ❤️ note to @evilsoft:**  In case you couldn't tell, this PR is me saying that I want to see more of what you can do in backend apps with `crocks`.  Also, you are specially positioned with <redacted> to really test the new ADT bits in this.  Please do that, because I don't want to ship this until I know it makes your life easier.  And take a look at my function signatures in the docs.  Things got weird when all the legacy bits were Promised-based, but then the signature for the `app` option was `Request -> m Response`.  Let me know how I can make that clearer if need be.
